linker/validate.go

```diff
@@ -76,3 +84,208 @@ func (r *result) validateExtension(fld protoreflect.FieldDescriptor, handler *re
+		return err
+	}
+
+	for _, nmd := range md.GetNestedType() {
+		if err := r.validateJSONNamesInMessage(nmd, handler); err != nil {
+			return err
+		}
+	}
+	for _, ed := range md.GetEnumType() {
+		if err := r.validateJSONNamesInEnum(ed, handler); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *result) validateJSONNamesInEnum(ed *descriptorpb.EnumDescriptorProto, handler *reporter.Handler) error {
+	seen := map[string]*descriptorpb.EnumValueDescriptorProto{}
+	for _, evd := range ed.GetValue() {
+		scope := "enum value " + ed.GetName() + "." + evd.GetName()
+
+		name := canonicalEnumValueName(evd.GetName(), ed.GetName())
+		if existing, ok := seen[name]; ok && evd.GetNumber() != existing.GetNumber() {
+			fldNode := r.EnumValueNode(evd)
+			existingNode := r.EnumValueNode(existing)
+			conflictErr := fmt.Errorf("%s: camel-case name (with optional enum name prefix removed) %q conflicts with camel-case name of enum value %s, defined at %v",
+				scope, name, existing.GetName(), r.FileNode().NodeInfo(existingNode).Start())
+
+			// Since proto2 did not originally have a JSON format, we report conflicts as just warnings
+			if r.Syntax() != protoreflect.Proto3 {
+				handler.HandleWarning(r.FileNode().NodeInfo(fldNode).Start(), conflictErr)
+			} else if err := handler.HandleErrorf(r.FileNode().NodeInfo(fldNode).Start(), conflictErr.Error()); err != nil {
+				return err
+			}
+		} else {
+			seen[name] = evd
+		}
+	}
+	return nil
+}
+
+func (r *result) validateFieldJSONNames(md *descriptorpb.DescriptorProto, useCustom bool, handler *reporter.Handler) error {
+	type jsonName struct {
+		source *descriptorpb.FieldDescriptorProto
+		// field's original JSON nane (which can differ in case from map key)
+		orig string
+		// true if orig is a custom JSON name (vs. the field's default JSON name)
+		custom bool
+	}
+	seen := map[string]jsonName{}
+
+	for _, fd := range md.GetField() {
+		scope := "field " + md.GetName() + "." + fd.GetName()
+		defaultName := internal.JSONName(fd.GetName())
+		name := defaultName
+		custom := false
+		if useCustom {
+			n := fd.GetJsonName()
+			if n != defaultName || r.hasCustomJSONName(fd) {
+				name = n
+				custom = true
+			}
+		}
+		lcaseName := strings.ToLower(name)
+		if existing, ok := seen[lcaseName]; ok {
+			// When useCustom is true, we'll only report an issue when a conflict is
+			// due to a custom name. That way, we don't double report conflicts on
+			// non-custom names.
+			if !useCustom || custom || existing.custom {
+				fldNode := r.FieldNode(fd)
+				customStr, srcCustomStr := "custom", "custom"
+				if !custom {
+					customStr = "default"
+				}
+				if !existing.custom {
+					srcCustomStr = "default"
+				}
+				otherName := ""
+				if name != existing.orig {
+					otherName = fmt.Sprintf(" %q", existing.orig)
+				}
+				conflictErr := fmt.Errorf("%s: %s JSON name %q conflicts with %s JSON name%s of field %s, defined at %v",
```

Comment:
Passing `Error()` string as a format argument to `HandleErrorf` is error-prone: it could include sequences that look like argument references.
```suggestion
				pos := r.FileNode().NodeInfo(fldNode).Start()
				conflictErr := reporter.Errorf(pos, "%s: %s JSON name %q conflicts with %s JSON name%s of field %s, defined at %v",
```